File: src/lib/pilot/services/knowledge.ts
import { getSql } from "@/lib/db";
import { ApiError } from "../errors";
import { rid, nextPublicId } from "../ids";
import { audit, ensureSeeded } from "../store";
import { chunkDocument } from "../rag/chunking";
import { SandboxEmbeddingProvider } from "../rag/embedding";
import { detectPromptInjection } from "./../rag/injection";
import { retrieve } from "../rag/retrieval";
import { rankEvidence } from "../rag/ranking";
import { extractDocument } from "../extraction";
import { sanitizeFilename, validateUpload } from "../safety";
import {
  DEMO_OPERATOR_ID,
  DEMO_WORKSPACE_ID,
  type DocumentStatus,
  type KnowledgeChunk,
  type KnowledgeDocument,
} from "../types";
import { iso, mapChunk, mapDoc } from "./tickets";

const embedder = new SandboxEmbeddingProvider();

export async function listDocuments(filters: {
  q?: string;
  category?: string;
  status?: DocumentStatus | "all";
}) {
  await ensureSeeded();
  const sql = await getSql();
  const rows = await sql.query<Record<string, unknown>>(
    `select * from knowledge_documents where workspace_id = $1 order by title`,
    [DEMO_WORKSPACE_ID],
  );
  let docs = rows.map(mapDoc);
  if (filters.q) {
    const q = filters.q.toLowerCase();
    docs = docs.filter(
      (d) =>
        d.title.toLowerCase().includes(q) ||
        d.publicId.toLowerCase().includes(q) ||
        d.filename.toLowerCase().includes(q),
    );
  }
  if (filters.category && filters.category !== "all") {
    docs = docs.filter((d) => d.category === filters.category);
  }
  if (filters.status && filters.status !== "all") {
    docs = docs.filter((d) => d.status === filters.status);
  }
  return docs;
}

export async function getDocument(id: string) {
  await ensureSeeded();
  const sql = await getSql();
  const rows = await sql.query<Record<string, unknown>>(
    `select * from knowledge_documents where workspace_id = $1 and (id = $2 or public_id = $2)`,
    [DEMO_WORKSPACE_ID, id],
  );
  if (!rows[0]) throw new ApiError("not_found", "Document not found.", 404);
  const doc = mapDoc(rows[0]);
  const chunks = await sql.query<Record<string, unknown>>(
    `select * from knowledge_chunks where document_id = $1 order by ordinal`,
    [doc.id],
  );
  const events = await sql.query<Record<string, unknown>>(
    `select * from document_events where document_id = $1 order by created_at asc`,
    [doc.id],
  );
  const versions = [
    {
      version: doc.version,
      status: doc.status,
      indexedAt: doc.lastIndexedAt,
      note: "Current indexed version",
    },
  ];
  return {
    document: doc,
    chunks: chunks.map(mapChunk),
    events: events.map((e) => ({
      id: String(e.id),
      stage: String(e.stage),
      detail: String(e.detail),
      createdAt: iso(e.created_at),
    })),
    versions,
  };
}

export async function createTextDocument(input: {
  title: string;
  body: string;
  category: string;
  filename?: string;
}) {
  if (!input.title.trim() || !input.body.trim()) {
    throw new ApiError("validation_error", "Title and body are required.", 422);
  }
  await ensureSeeded();
  const sql = await getSql();
  const count = await sql.query<{ n: number }>(`select count(*)::int as n from knowledge_documents`);
  const id = rid("doc");
  const publicId = nextPublicId("DOC", (count[0]?.n ?? 0) + 21);
  const filename = sanitizeFilename(input.filename || `${slug(input.title)}.md`);
  await sql.query(
    `insert into knowledge_documents (
      id, workspace_id, public_id, title, category, version, source_type, status,
      chunk_count, last_indexed_at, owner_id, owner_name, usage_count, filename, body, failed_reason, suspicious
    ) values ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17)`,
    [
      id,
      DEMO_WORKSPACE_ID,
      publicId,
      input.title.trim(),
      input.category || "policy",
      "1.0",
      "markdown",
      "pending",
      0,
      null,
      DEMO_OPERATOR_ID,
      "Elena Voss",
      0,
      filename,
      input.body,
      null,
      0,
    ],
  );
  await addEvent(id, "pending", "Document accepted");
  await indexDocument(id);
  await audit("document.created", "document", id, { publicId });
  return getDocument(id);
}

export async function uploadDocument(filename: string, mime: string, content: string, size: number) {
  const err = validateUpload(filename, mime, size);
  if (err) throw new ApiError("invalid_upload", err, 422);
  const extracted = await extractDocument(filename, content, mime);
  return createTextDocument({
    title: filename.replace(/\.[^.]+$/, "").replace(/[_-]+/g, " "),
    body: extracted.text,
    category: "policy",
    filename,
  });
}

export async function reindexDocument(id: string) {
  const doc = await getDocument(id);
  if (doc.document.status === "archived") {
    throw new ApiError("invalid_state", "Archived documents cannot be reindexed.", 409);
  }
  await indexDocument(doc.document.id);
  await audit("document.reindexed", "document", doc.document.id, {});
  return getDocument(doc.document.id);
}

export async function archiveDocument(id: string) {
  const doc = await getDocument(id);
  const sql = await getSql();
  await sql.query(`update knowledge_documents set status = 'archived' where id = $1 and workspace_id = $2`, [
    doc.document.id,
    DEMO_WORKSPACE_ID,
  ]);
  await addEvent(doc.document.id, "archived", "Archived by operator");
  await audit("document.archived", "document", doc.document.id, {});
  return getDocument(doc.document.id);
}

export async function testRetrieval(question: string) {
  if (!question.trim()) throw new ApiError("validation_error", "Enter a question to test retrieval.", 422);
  await ensureSeeded();
  const sql = await getSql();
  const docs = await sql.query<Record<string, unknown>>(
    `select * from knowledge_documents where workspace_id = $1 and status = 'ready'`,
    [DEMO_WORKSPACE_ID],
  );
  const chunks = await sql.query<Record<string, unknown>>(
    `select * from knowledge_chunks where workspace_id = $1`,
    [DEMO_WORKSPACE_ID],
  );
  const mappedDocs = docs.map(mapDoc);
  const mappedChunks = chunks.map(mapChunk).filter((c) => !c.flagged);
  const titleByDoc = new Map(mappedDocs.map((d) => [d.id, d.title]));
  const hits = rankEvidence(retrieve(question, mappedChunks, embedder, 8), titleByDoc, question);
  const runId = rid("run");
  await sql.query(
    `insert into retrieval_runs (id, workspace_id, ticket_id, query, created_at) values ($1,$2,$3,$4,$5)`,
    [runId, DEMO_WORKSPACE_ID, null, question, new Date().toISOString()],
  );
  return {
    query: question,
    runId,
    results: hits.map((h) => ({
      chunkId: h.chunk.id,
      documentId: h.chunk.documentId,
      documentTitle: titleByDoc.get(h.chunk.documentId) ?? "Unknown",
      section: h.chunk.section,
      excerpt: h.chunk.text.slice(0, 500),
      score: Math.round(h.score * 1000) / 1000,
      cosine: Math.round(h.cosine * 1000) / 1000,
      lexical: Math.round(h.lexical * 1000) / 1000,
      flagged: h.chunk.flagged,
    })),
  };
}

async function indexDocument(id: string) {
  const sql = await getSql();
  const rows = await sql.query<Record<string, unknown>>(
    `select * from knowledge_documents where id = $1 and workspace_id = $2`,
    [id, DEMO_WORKSPACE_ID],
  );
  const doc = rows[0];
  if (!doc) throw new ApiError("not_found", "Document not found.", 404);
  if (!String(doc.body ?? "").trim()) {
    await sql.query(
      `update knowledge_documents set status = 'failed', failed_reason = $1 where id = $2`,
      ["No extractable text.", id],
    );
    await addEvent(id, "failed", "No extractable text");
    return;
  }
  await sql.query(`update knowledge_documents set status = 'extracting' where id = $1`, [id]);
  await addEvent(id, "extracting", "Text extracted");
  await sql.query(`update knowledge_documents set status = 'chunking' where id = $1`, [id]);
  const chunks = chunkDocument(String(doc.body));
  await addEvent(id, "chunking", `${chunks.length} chunks`);
  await sql.query(`update knowledge_documents set status = 'embedding' where id = $1`, [id]);
  await sql.query(`delete from knowledge_chunks where document_id = $1`, [id]);
  let flagged = 0;
  for (const chunk of chunks) {
    const inj = detectPromptInjection(chunk.text);
    if (inj.matched) flagged += 1;
    await sql.query(
      `insert into knowledge_chunks (
        id, document_id, workspace_id, ordinal, section, page, text, embedding, token_count, flagged, flag_reason
      ) values ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11)`,
      [
        rid("chk"),
        id,
        DEMO_WORKSPACE_ID,
        chunk.ordinal,
        chunk.section,
        chunk.page,
        chunk.text,
        JSON.stringify(embedder.embed(chunk.text)),
        chunk.tokenCount,
        inj.matched ? 1 : 0,
        inj.matched ? `Prompt-injection patterns: ${inj.patterns.join(", ")}` : null,
      ],
    );
  }
  const now = new Date().toISOString();
  await sql.query(
    `update knowledge_documents set status = 'ready', chunk_count = $1, last_indexed_at = $2, failed_reason = null where id = $3`,
    [chunks.length, now, id],
  );
  await addEvent(
    id,
    "ready",
    flagged ? `Indexed with ${flagged} flagged chunk(s) excluded from answers` : "Indexed",
  );
}

async function addEvent(documentId: string, stage: string, detail: string) {
  const sql = await getSql();
  await sql.query(
    `insert into document_events (id, document_id, stage, detail, created_at) values ($1,$2,$3,$4,$5)`,
    [rid("dov"), documentId, stage, detail, new Date().toISOString()],
  );
}

function slug(s: string): string {
  return s.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "").slice(0, 60);
}

export type { KnowledgeDocument, KnowledgeChunk };

File: backend/app/services/knowledge.py
"""Knowledge document ingest, listing, reindex, and retrieval testing."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any

from sqlalchemy import func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.errors import ApiError
from app.core.ids import next_public_id, rid
from app.core.safety import sanitize_filename, validate_upload
from app.models import DocumentEvent, KnowledgeChunk, KnowledgeDocument, RetrievalRun
from app.providers.sandbox import SandboxEmbeddingProvider
from app.rag.chunking import chunk_document
from app.rag.extraction import extract_document
from app.rag.injection import detect_prompt_injection
from app.rag.ranking import rank_evidence
from app.rag.retrieval import retrieve
from app.seed.people import DEMO_OPERATOR_ID
from app.services.audit import write_audit
from app.services.serialize import iso, map_chunk, map_document

embedder = SandboxEmbeddingProvider()


def _slug(title: str) -> str:
    return "".join(ch.lower() if ch.isalnum() else "-" for ch in title).strip("-")[:80] or "document"


async def get_document_row(session: AsyncSession, workspace_id: str, document_id: str) -> KnowledgeDocument:
    result = await session.execute(
        select(KnowledgeDocument).where(
            KnowledgeDocument.workspace_id == workspace_id,
            or_(KnowledgeDocument.id == document_id, KnowledgeDocument.public_id == document_id),
        )
    )
    doc = result.scalar_one_or_none()
    if doc is None:
        raise ApiError("not_found", "Document not found.", 404)
    return doc


async def list_documents(
    session: AsyncSession,
    workspace_id: str,
    filters: dict[str, Any] | None = None,
) -> list[dict[str, Any]]:
    rows = (
        (
            await session.execute(
                select(KnowledgeDocument)
                .where(KnowledgeDocument.workspace_id == workspace_id)
                .order_by(KnowledgeDocument.title)
            )
        )
        .scalars()
        .all()
    )
    docs = [map_document(d) for d in rows]
    filters = filters or {}
    q = (filters.get("q") or "").lower()
    if q:
        docs = [d for d in docs if q in d["title"].lower() or q in d["public_id"].lower() or q in d["filename"].lower()]
    if filters.get("category") and filters["category"] != "all":
        docs = [d for d in docs if d["category"] == filters["category"]]
    if filters.get("status") and filters["status"] != "all":
        docs = [d for d in docs if d["status"] == filters["status"]]
    return docs


async def get_document(session: AsyncSession, workspace_id: str, document_id: str) -> dict[str, Any]:
    doc = await get_document_row(session, workspace_id, document_id)
    chunks = (
        (
            await session.execute(
                select(KnowledgeChunk).where(KnowledgeChunk.document_id == doc.id).order_by(KnowledgeChunk.ordinal)
            )
        )
        .scalars()
        .all()
    )
    events = (
        (
            await session.execute(
                select(DocumentEvent)
                .where(DocumentEvent.document_id == doc.id)
                .order_by(DocumentEvent.created_at.asc())
            )
        )
        .scalars()
        .all()
    )
    mapped = map_document(doc)
    return {
        "document": mapped,
        "chunks": [map_chunk(c) for c in chunks],
        "events": [{"id": e.id, "stage": e.stage, "detail": e.detail, "created_at": iso(e.created_at)} for e in events],
        "versions": [
            {
                "version": mapped["version"],
                "status": mapped["status"],
                "indexed_at": mapped["last_indexed_at"],
                "note": "Current indexed version",
            }
        ],
    }


async def add_event(session: AsyncSession, document_id: str, stage: str, detail: str) -> None:
    session.add(
        DocumentEvent(
            id=rid("dov"),
            document_id=document_id,
            stage=stage,
            detail=detail,
            created_at=datetime.now(timezone.utc),
        )
    )


async def index_document(session: AsyncSession, document_id: str) -> None:
    doc = await session.get(KnowledgeDocument, document_id)
    if doc is None:
        raise ApiError("not_found", "Document not found.", 404)
    now = datetime.now(timezone.utc)
    existing = (
        (await session.execute(select(KnowledgeChunk).where(KnowledgeChunk.document_id == document_id))).scalars().all()
    )
    for existing_chunk in existing:
        await session.delete(existing_chunk)
    await add_event(session, document_id, "extracting", "Extracting text")
    await add_event(session, document_id, "chunking", "Splitting into overlapping passages")
    pieces = chunk_document(doc.body) if doc.body else []
    await add_event(session, document_id, "embedding", "Generating sandbox hash embeddings")
    flagged_any = False
    for piece in pieces:
        inj = detect_prompt_injection(piece.text)
        flagged_any = flagged_any or inj.matched
        session.add(
            KnowledgeChunk(
                id=rid("chk"),
                document_id=doc.id,
                workspace_id=doc.workspace_id,
                ordinal=piece.ordinal,
                section=piece.section,
                page=piece.page,
                text=piece.text,
                embedding=json.dumps(embedder.embed(piece.text)),
                token_count=piece.token_count,
                flagged=inj.matched,
                flag_reason=(f"Prompt-injection patterns: {', '.join(inj.patterns)}" if inj.matched else None),
            )
        )
    doc.status = "ready"
    doc.chunk_count = len(pieces)
    doc.last_indexed_at = now
    doc.suspicious = flagged_any or bool(doc.suspicious)
    await add_event(
        session,
        document_id,
        "ready",
        "Indexed by sandbox embedding provider"
        + ("; injection patterns flagged and excluded from retrieval" if flagged_any else ""),
    )


async def create_text_document(
    session: AsyncSession,
    workspace_id: str,
    input_data: dict[str, Any],
    actor: dict[str, str],
) -> dict[str, Any]:
    title = (input_data.get("title") or "").strip()
    body = (input_data.get("body") or "").strip()
    if not title or not body:
        raise ApiError("validation_error", "Title and body are required.", 422)
    count = (await session.execute(select(func.count()).select_from(KnowledgeDocument))).scalar_one()
    doc_id = rid("doc")
    public_id = next_public_id("DOC", int(count or 0) + 21)
    filename = sanitize_filename(input_data.get("filename") or f"{_slug(title)}.md")
    session.add(
        KnowledgeDocument(
            id=doc_id,
            workspace_id=workspace_id,
            public_id=public_id,
            title=title,
            category=input_data.get("category") or "policy",
            version="1.0",
            source_type="markdown",
            status="pending",
            chunk_count=0,
            last_indexed_at=None,
            owner_id=actor.get("id") or DEMO_OPERATOR_ID,
            owner_name=actor.get("name") or "Elena Voss",
            usage_count=0,
            filename=filename,
            body=body,
            failed_reason=None,
            suspicious=False,
        )
    )
    await add_event(session, doc_id, "pending", "Document accepted")
    await session.flush()
    await index_document(session, doc_id)
    await write_audit(
        session,
        "document.created",
        "document",
        doc_id,
        {"publicId": public_id},
        workspace_id=workspace_id,
        actor_id=actor["id"],
        actor_name=actor["name"],
    )
    await session.flush()
    return await get_document(session, workspace_id, doc_id)


async def upload_document(
    session: AsyncSession,
    workspace_id: str,
    filename: str,
    mime: str,
    content: str | bytes,
    size: int,
    actor: dict[str, str],
) -> dict[str, Any]:
    err = validate_upload(filename, mime, size)
    if err:
        raise ApiError("invalid_upload", err, 422)
    extracted = extract_document(filename, content, mime)
    return await create_text_document(
        session,
        workspace_id,
        {
            "title": filename.rsplit(".", 1)[0].replace("_", " ").replace("-", " "),
            "body": extracted.text,
            "category": "policy",
            "filename": filename,
        },
        actor,
    )


async def reindex_document(
    session: AsyncSession, workspace_id: str, document_id: str, actor: dict[str, str]
) -> dict[str, Any]:
    doc = await get_document_row(session, workspace_id, document_id)
    if doc.status == "archived":
        raise ApiError("invalid_state", "Archived documents cannot be reindexed.", 409)
    if doc.status == "failed" and not doc.body:
        raise ApiError("invalid_state", "Failed ingest has no extractable text to reindex.", 409)
    await index_document(session, doc.id)
    await write_audit(
        session,
        "document.reindexed",
        "document",
        doc.id,
        {},
        workspace_id=workspace_id,
        actor_id=actor["id"],
        actor_name=actor["name"],
    )
    await session.flush()
    return await get_document(session, workspace_id, doc.id)


async def archive_document(
    session: AsyncSession, workspace_id: str, document_id: str, actor: dict[str, str]
) -> dict[str, Any]:
    doc = await get_document_row(session, workspace_id, document_id)
    doc.status = "archived"
    await add_event(session, doc.id, "archived", "Archived by operator")
    await write_audit(
        session,
        "document.archived",
        "document",
        doc.id,
        {},
        workspace_id=workspace_id,
        actor_id=actor["id"],
        actor_name=actor["name"],
    )
    await session.flush()
    return await get_document(session, workspace_id, doc.id)


async def test_retrieval(session: AsyncSession, workspace_id: str, question: str) -> dict[str, Any]:
    if not question.strip():
        raise ApiError("validation_error", "Enter a question to test retrieval.", 422)
    docs = (
        (
            await session.execute(
                select(KnowledgeDocument).where(
                    KnowledgeDocument.workspace_id == workspace_id, KnowledgeDocument.status == "ready"
                )
            )
        )
        .scalars()
        .all()
    )
    chunks = (
        (await session.execute(select(KnowledgeChunk).where(KnowledgeChunk.workspace_id == workspace_id)))
        .scalars()
        .all()
    )
    mapped_docs = [map_document(d) for d in docs]
    mapped_chunks = [map_chunk(c) for c in chunks if not c.flagged]
    title_by_doc = {d["id"]: d["title"] for d in mapped_docs}
    hits = rank_evidence(retrieve(question, mapped_chunks, embedder, 8), title_by_doc, question)
    run_id = rid("run")
    session.add(
        RetrievalRun(
            id=run_id,
            workspace_id=workspace_id,
            ticket_id=None,
            query=question,
            created_at=datetime.now(timezone.utc),
        )
    )
    await session.flush()
    return {
        "query": question,
        "run_id": run_id,
        "results": [
            {
                "chunk_id": h["chunk"]["id"] if isinstance(h["chunk"], dict) else h["chunk"].id,
                "document_id": h["chunk"]["document_id"] if isinstance(h["chunk"], dict) else h["chunk"].document_id,
                "document_title": title_by_doc.get(
                    h["chunk"]["document_id"] if isinstance(h["chunk"], dict) else h["chunk"].document_id,
                    "Unknown",
                ),
                "section": h["chunk"]["section"] if isinstance(h["chunk"], dict) else h["chunk"].section,
                "score": float(h["score"]),
                "cosine": float(h.get("cosine") or 0),
                "lexical": float(h.get("lexical") or 0),
                "excerpt": (h["chunk"]["text"] if isinstance(h["chunk"], dict) else h["chunk"].text)[:420],
            }
            for h in hits
        ],
    }
